### PR TITLE
Hoverformat bug with showTickLabels

### LINKED ISSUE #261: Hoverformat fails to apply when showticklabels=False
This is in reference to the following issue in the Plotly Python API: 
[Issue 398](https://github.com/plotly/plotly.py/issues/398)

Unless `showticklabels = True` then the hoverformat is not applied. 
For example in the following code:

```
from plotly.graph_objs import Scatter, Layout, YAxis, Figure
import plotly.plotly as py

data = [Scatter(x=[1, 2, 3],
                 y=[0.12345, 0.23456, 0.34567])]
layout = Layout(yaxis=YAxis(showticklabels=False,
                            hoverformat=",.2r",))
figure = Figure(data=data, layout=layout)
py.iplot(figure)
```
[Plot generated from the code](http://codepen.io/etpinard/pen/jWRzrR)

The associated yaxis portion of the json is
```
yaxis:{
  hoverformat:",.2r",
  showticklabels:false,
  type:"linear",
  range:[
             -0.2711700793681046,
              0.7894967832778617
             ],
  autorange:false
}
```
and so it appears to have translated correctly from the python plot function call.

### PR DESCRIPTION
**Pull request by mdtusz**
Fixes #261.

Pulled some of the coersions out from the block that checks for `showTickLabels` so things act as expected, and added a quick testeroonie in there for good measure in the future.

### CODE REVIEW

File: test/jasmine/tests/hover_label_test.js [hunk 1]
```diff
@@ -242,4 +243,42 @@ describe('hover info', function() {
...
+                y: [0.12345, 0.23456, 0.34567]
+            }],
+            layout = {
+                yaxis: { showticklabels: true, hoverformat: ',.2r' },
+                width: 600,
+                height: 400
+            };
+
+        beforeEach(function() {
+            this. gd = createGraphDiv();
```
Comment (etpinard) [inline test/jasmine/tests/hover_label_test.js]: Haha. Look at that. A jasmine `this`. Maybe I should start using those more.

File: test/jasmine/tests/hover_label_test.js [hunk 2]
```diff
@@ -242,4 +243,42 @@ describe('hover info', function() {
...
+
+        it('should display the correct format when ticklabels false', function() {
+            layout.yaxis.showticklabels = false;
+            Plotly.plot(this.gd, data, layout);
+            mouseEvent('mousemove', 310, 220);
+
+            var hovers = d3.selectAll('g.hovertext');
+
+            expect(hovers.size()).toEqual(1);
+            expect(hovers.select('text')[0][0].textContent).toEqual('0.23');
```
Comment (etpinard) [inline test/jasmine/tests/hover_label_test.js]: dope tests.

File: test/jasmine/tests/hover_label_test.js (hunk 1 above)
Comment (mdtusz) [inline test/jasmine/tests/hover_label_test.js]: :smile_cat: I'm ok with `this` when it's idiomatic.

File: src/plots/cartesian/tick_defaults.js [hunk 3]
```diff
@@ -29,28 +29,28 @@ module.exports = function handleTickDefaults(containerIn, containerOut, coerce,
...
+    if(axType !== 'category') {
+        if(!options.noHover) coerce('hoverformat');
 
-            if(!tickFormat && axType !== 'date') {
-                coerce('showexponent', showAttrDflt);
-                coerce('exponentformat');
-            }
+        if(!tickFormat && axType !== 'date') {
+            coerce('showexponent', showAttrDflt);
+            coerce('exponentformat');
```
Comment (etpinard) [inline src/plots/cartesian/tick_defaults.js]: Does the exponent format affect hover labels? 

Or does it only affect the tick labels?

If the latter, we should bring it back into the `if(showTickLabels)` block.

File: src/plots/cartesian/tick_defaults.js [hunk 4]
```diff
@@ -29,28 +29,28 @@ module.exports = function handleTickDefaults(containerIn, containerOut, coerce,
         delete containerOut.tickcolor;
     }
 
-    var showTickLabels = coerce('showticklabels');
+    var showAttrDflt = getShowAttrDflt(containerIn);
+
+    var showTickLabels = coerce('showticklabels'),
+        tickFormat = coerce('tickformat'),
```
Comment (etpinard) [inline src/plots/cartesian/tick_defaults.js]: similarly here.

**Comment by etpinard** [comment]
Great.

Once you figured out if `tickformat` and `exponentformat` can affect the hover labels, could you add some jasmine tests for `handleTickDefaults` in https://github.com/plotly/plotly.js/blob/master/test/jasmine/tests/axes_test.js ?

**Comment by mdtusz** [comment]
@etpinard let's try to get this one in for the next release too if we can.

File: src/plots/cartesian/tick_defaults.js [hunk 5]
```diff
@@ -29,29 +29,27 @@ module.exports = function handleTickDefaults(containerIn, containerOut, coerce,
...
-                coerce('showexponent', showAttrDflt);
-                coerce('exponentformat');
-            }
-        }
-
         var tickPrefix = coerce('tickprefix');
         if(tickPrefix) coerce('showtickprefix', showAttrDflt);
 
         var tickSuffix = coerce('ticksuffix');
         if(tickSuffix) coerce('showticksuffix', showAttrDflt);
```
Comment (etpinard) [inline src/plots/cartesian/tick_defaults.js]: These are not quite right still.

http://codepen.io/etpinard/pen/MKMNXe

`tickprefix` and `ticksuffix` have an effect (and should) of the hover label; they should be coerced regardless of the `showTickLabels`  value.

Comment (mdtusz) [inline src/plots/cartesian/tick_defaults.js]: Ah yeah that's what I had done before but I thought you were saying they *should* be in the `showTickLabels` block. I'll pull them out again then.

Comment (etpinard) [inline src/plots/cartesian/tick_defaults.js]: Yeah, my bad. I wasn't very clear in my previous comment.

File: src/plots/cartesian/tick_defaults.js [hunk 6]
```diff
@@ -29,29 +29,27 @@ module.exports = function handleTickDefaults(containerIn, containerOut, coerce,
...
-        if(axType !== 'category') {
-            var tickFormat = coerce('tickformat');
-            if(!options.noHover) coerce('hoverformat');
-
-            if(!tickFormat && axType !== 'date') {
-                coerce('showexponent', showAttrDflt);
-                coerce('exponentformat');
-            }
+        var tickFormat = coerce('tickformat');
+        if(!tickFormat && axType !== 'date') {
```
Comment (etpinard) [inline src/plots/cartesian/tick_defaults.js]: why is this block not wrapped inside `axType !== 'category'` ?

Comment (mdtusz) [inline src/plots/cartesian/tick_defaults.js]: Whoa good catch. When I pulled out `coerce('hoverformat')` I completely goofed on that. I'll add it into the `if` condition right now.

**Comment by etpinard** [comment]
yeah man :dancer:

**Pull request merged**